rustls/src/dns_name.rs

```diff
@@ -0,0 +1,176 @@
+                    return Err(InvalidDnsNameError);
+                }
+                label_length = 0;
+            }
+            // first character of label must be <letter>
+            b'a'..=b'z' | b'A'..=b'Z' if label_length == 0 => {
+                previous_label_ch = *ch;
+                label_length += 1;
+            }
+            // subsequent characters can be <letter> | <digit> | "-" | "_"
+            // (underscore is a relaxation of RFC1035)
+            b'a'..=b'z' | b'A'..=b'Z' | b'-' | b'_' | b'0'..=b'9' if label_length > 0 => {
+                previous_label_ch = *ch;
+                label_length += 1;
+            }
+            // other characters are illegal
+            _ => {
+                return Err(InvalidDnsNameError);
+            }
+        };
+
+        if label_length > MAX_LABEL_LENGTH {
+            return Err(InvalidDnsNameError);
+        }
+    }
+
+    if previous_label_ch == b'-' {
+        return Err(InvalidDnsNameError);
+    }
+
+    Ok(())
+}
+
+/// The provided input could not be parsed because
+/// it is not a syntactically-valid DNS Name.
+#[derive(Debug)]
+pub struct InvalidDnsNameError;
+
+impl fmt::Display for InvalidDnsNameError {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        f.write_str("invalid dns name")
+    }
+}
+
+impl StdError for InvalidDnsNameError {}
+
+/// A type which encapsulates an owned string that is a syntactically valid DNS name.
+#[derive(Clone, Eq, Hash, PartialEq, Debug)]
+pub struct DnsName(String);
+
+impl<'a> DnsName {
+    /// Produce a borrowed `DnsNameRef` from this owned `DnsName`.
+    pub fn borrow(&'a self) -> DnsNameRef<'a> {
+        DnsNameRef(self.as_ref())
+    }
+
+    /// Validate the given bytes are a DNS name if they are viewed as ASCII.
+    pub fn try_from_ascii(bytes: &[u8]) -> Result<Self, InvalidDnsNameError> {
+        // nb. a sequence of bytes that is accepted by `validate()` is both
+        // valid UTF-8, and valid ASCII.
+        String::from_utf8(bytes.to_vec())
+            .map_err(|_| InvalidDnsNameError)
+            .and_then(Self::try_from)
+    }
+}
+
+impl TryFrom<String> for DnsName {
+    type Error = InvalidDnsNameError;
+
+    fn try_from(value: String) -> Result<Self, Self::Error> {
+        validate(value.as_bytes())?;
+        Ok(Self(value))
+    }
+}
+
+impl AsRef<str> for DnsName {
+    fn as_ref(&self) -> &str {
+        AsRef::<str>::as_ref(&self.0)
+    }
+}
+
+/// A type which encapsulates a borrowed string that is a syntactically valid DNS name.
+#[derive(Eq, Hash, PartialEq, Debug)]
+pub struct DnsNameRef<'a>(&'a str);
+
+impl<'a> DnsNameRef<'a> {
+    /// Copy this object to produce an owned `DnsName`.
+    pub fn to_owned(&'a self) -> DnsName {
+        DnsName(self.0.to_string())
+    }
+
+    /// Copy this object to produce an owned `DnsName`, smashing the case to lowercase
+    /// in one operation.
+    pub fn to_owned_normalised(&'a self) -> DnsName {
```

Comment:
I think we should use US spelling in API?

Reply:
-> `to_lowercase_owned` to avoid the question and better express the purpose.